lib/globus_client/endpoint.rb

```diff
@@ -98,20 +94,22 @@ def path_segments
     end
 
     # @param filepath [String] an absolute path to look up contents e.g. /uploads/example/work123/version1
-    # @param filenames [Array<Hash>] an array of Hashes, with keys for filename with path and filesize
-    def ls_path(filepath, filenames)
+    # @param files [Array<FileInfo>] an array of FileInfo structs, each of which has a name and a size
+    def ls_path(filepath, files)
       # List files recursively at an endpoint https://docs.globus.org/api/transfer/file_operations/#list_directory_contents
       response = connection.get("#{transfer_path}/ls?path=#{filepath}")
       if response.success?
         data = JSON.parse(response.body)["DATA"]
-        data.select { |object| object["type"] == "file" }.map { |file| file }
-          .each { |file| filenames << {filename: "#{filepath}#{file["name"]}", filesize: file["size"]} }
-        data.select { |object| object["type"] == "dir" }.map { |dir| dir["name"] }
```

Comment:
Got rid of the `#map` and put that "we care about only the name of the dir" jazz in the `#each` block to avoid iterating where we might not need to.